consensus/hotstuff/votecollector/consensus_cluster_vote_collector.go

```diff
@@ -0,0 +1,120 @@
+package votecollector
+
+import (
+	"fmt"
+
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/consensus/hotstuff"
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type ConsensusClusterVoteCollector struct {
+	BaseVoteCollector
+
+	dkg           hotstuff.DKG
+	aggregator    CombinedAggregator
+	reconstructor RandomBeaconReconstructor
+	onQCCreated   hotstuff.OnQCCreated
+	done          atomic.Bool
+}
+
+func NewVerifyingVoteCollector(base BaseVoteCollector) *ConsensusClusterVoteCollector {
+	return &ConsensusClusterVoteCollector{
+		BaseVoteCollector: base,
+	}
+}
+
+// CreateVote implements BlockSigner interface for creating votes from block proposals
+func (c *ConsensusClusterVoteCollector) CreateVote(block *model.Block) (*model.Vote, error) {
+	panic("implement me")
+}
+
+func (c *ConsensusClusterVoteCollector) AddVote(vote *model.Vote) error {
+	if c.done.Load() {
+		return nil
+	}
+
+	verified, sigType, err := c.aggregator.Verify(vote.SignerID, vote.SigData)
+	if err != nil {
+		return fmt.Errorf("could not verify vote signature: %w", err)
+	}
+
+	// TODO: handle if verified == false
+	if !verified {
+		return fmt.Errorf("could not verify vote signature: %w", err)
+	}
+
+	if c.done.Load() {
+		return nil
+	}
+
+	_, err = c.aggregator.TrustedAdd(vote.SignerID, vote.SigData, sigType)
+	if err != nil {
+		return fmt.Errorf("could not aggregate vote signature: %w", err)
+	}
+
+	if sigType == SigTypeThreshold {
+		index, err := c.dkg.Index(vote.SignerID)
+		if err != nil {
+			return fmt.Errorf("could not retrieve dkg index for signer (%v): %w", vote.SignerID, err)
+		}
+		_, err = c.reconstructor.TrustedAdd(index, vote.SigData)
+		if err != nil {
+			return fmt.Errorf("could not add random beacon sig share: %w", err)
+		}
+	}
+
+	// we haven't collected sufficient weight, we have nothing to do further
+	if !c.aggregator.HasSufficientWeight() {
+		return nil
+	}
+
+	// we haven't collected sufficient shares, we have nothing to do further
+	if !c.reconstructor.HasSufficientShares() {
+		return nil
+	}
```

Comment:
what about a boolean or?

Reply:
Works too, basically took what was written in [notion](https://www.notion.so/dapperlabs/Architecture-Proposal-07c8a782e95c48cea67fa8caa1ea0e18).